### DISCUSSION TITLE
Problems with plotting data in RADOLAN (polar stereographic projection)

### DISCUSSION THREAD

**Discussion by ZofiaBaldysz** [dtcenter/METplus#1840] [File I/O] [answered]
Hi,
I have a NetCDF file with precipitation data saved in RADOLAN Grid (Polar Stereographic Projection) and I have difficulties plotting these data using plot_data_plane tool. 
By inspecting a NetCDF CF documentation (https://cfconventions.org/cf-conventions/cf-conventions.html#polar-stereographic) I found that necessary projection elements for Polar Stereographic are either standard_parallel (EPSG 9829) or scale_factor_at_projection_origin (EPSG 9810). 
However , when I’m using scale_factor from https://docs.wradlib.org/en/stable/notebooks/radolan/radolan_grid.html , (which according to NetCDF CF documentation should be ok, because it is bigger than 0, scale_factor_at_projection_origin > 0.0) I got information about unexpected value:

`NcCfFile::get_grid_mapping_polar_stereographic() -> Unexpected attribute value of 0.9 for the scale_factor_at_projection_origin attribute of the RADOLAN_grid variable.`

When I’m creating NetCDF file without scale_factor_at_projection_origin, and instead of this I’m setting standard_paralell, the METPlus gives me error:

`ERROR  : get_nc_var_att_double() -> Cannot get "scale_factor_at_projection_origin" from RADOLAN_grid variable.`

Is there any possibility to add such variables/attributes of polar stereographic projection to the netCDF file, to make it readable by METPlus?

Thanks,
Zofia

ncdump of my file:
```
dimensions:
        time = 864 ;
        y = 1100 ;
        x = 900 ;
variables:
        int time(time) ;
                time:standard_name = "time" ;
                time:long_name = "time minutes since 2021-07-13T00:00:00.000000000" ;
                time:units = "minutes since 2021-07-13T00:00:00.000000000" ;
        int RADOLAN_grid ;
                RADOLAN_grid:grid_mapping_name = "polar_stereographic" ;
                RADOLAN_grid:projection = "polar_stereographic" ;
                RADOLAN_grid:straight_vertical_longitude_from_pole = 10 ;
                RADOLAN_grid:latitude_of_projection_origin = 90 ;
                RADOLAN_grid:longitude_of_projection_origin = 10 ;
                RADOLAN_grid:horizontal_datum_name = "Radolan_Kugel" ;
                RADOLAN_grid:standard_parallel = 10 ;
                RADOLAN_grid:false_easting = 0 ;
                RADOLAN_grid:false_northing = 0 ;
                RADOLAN_grid:semi_major_axis = 6370040 ;
                RADOLAN_grid:semi_minor_axis = 6370040 ;
                RADOLAN_grid:crs_wkt = "PROJCS[\"Radolan Projection\",\n    GEOGCS[\"Radolan Coordinate System\",\n        DATUM[\"Radolan_Kugel\",\n            SPHEROID[\"Erdkugel\",6370040,0]],\n        PRIMEM[\"Greenwich\",0,\n            AUTHORITY[\"EPSG\",\"8901\"]],\n        UNIT[\"degree\",0.0174532925199433,\n            AUTHORITY[\"EPSG\",\"9122\"]]],\n    PROJECTION[\"Polar_Stereographic\"],\n    PARAMETER[\"latitude_of_origin\",60],\n    PARAMETER[\"central_meridian\",10],\n    PARAMETER[\"false_easting\",0],\n    PARAMETER[\"false_northing\",0],\n    UNIT[\"kilometre\",1000,\n        AUTHORITY[\"EPSG\",\"9036\"]],\n    AXIS[\"Easting\",SOUTH],\n    AXIS[\"Northing\",SOUTH]]" ;
        double lat(y, x) ;
                lat:standard_name = "latitude" ;
                lat:long_name = "latitude" ;
                lat:units = "degrees_north" ;
                lat:coorindates = "lat lon" ;
        double lon(y, x) ;
                lon:standard_name = "longitude" ;
                lon:long_name = "longitude" ;
                lon:units = "degrees_east" ;
                lon:coordinates = "lat lon" ;
        double y(y) ;
                y:standard_name = "projection_y_coordinate" ;
                y:long_name = "RADOLAN Grid y coordinate of projection" ;
                y:units = "km" ;
        double x(x) ;
                x:standard_name = "projection_x_coordinate" ;
                x:long_name = "RADOLAN Grid y coordinate of projection" ;
                x:units = "km" ;
        float PREC(time, y, x) ;
                PREC:_FillValue = NaNf ;
                PREC:standard_name = "rainfall_amount" ;
                PREC:long_name = "rainfall amount" ;
                PREC:description = "accumulated rainfall amount" ;
                PREC:units = "kg m-2" ;
                PREC:grid_mapping = "RADOLAN_grid" ;
                PREC:coordinates = "lat lon" ;

// global attributes:
                :TITLE = "Radar rainfall" ;
                :Conventions = "CF-1.7" ;
```

**Comment by CPKalb** [comment]
Hi Zofia:

We recently discovered a bug in the way MET parses Polar Stereographic grids from CF-compliant NetCDF files.  The bug is originally described [here](https://github.com/dtcenter/METplus/discussions/1718).  Specifically, there were two issues:

1.  longitude_of_projection_origin is NOT required.
2.  Either scale_factor_at_projection_origin or standard_parallel may be specified, but MET is not parsing/interpreting the latter.

Your second error is due to number 2.  It should be reading either scale_factor_at_projection_origin or standard_parallel, but it is not parsing/interpreting the standard_parallel.  We have a bugfix issue (https://github.com/dtcenter/MET/issues/2218) in MET to correct this, but it has not been merged into MET yet.  Please feel free to comment directly on the issue to let us know if/when it's needed.

I'm not familiar with how MET uses the scale_factor_at_projection_origin, so would need to check with others here, if that is needed as well.

Christina

**Reply by ZofiaBaldysz** [reply]
Hi Christina,
thank you very much for quick response. I think the problem is not exactly the same. If I provide only standard_parallel the MET requires also scale_factor_at_projection_origin. But, if provide only scale_factor_at_projection_origin I get the error message about wrong value. I found that if I provide value 1, it works, but then it is different projection. According to the projection details, the scale factor should be 0.93301270189 but there is no way how to put it into MET. It seems, that MET doesn't allow to provide the value below 1 (just my insights), while NETCDF CF standard allows these values to be above 0.

Zofia

**Reply by CPKalb** [reply]
Hi Zofia:

Thanks for clarifying.  MET should not require both the standard_parallel and scale_factor_at_projection_origin.  It should only require one or the other. 

Do you have a sample data file that could be [uploaded ](https://github.com/dtcenter/METplus/discussions/954) to our ftp.  It would be helpful for us to have a sample file that we could use to test the changes/enhancements that we are making to MET for polar stereographic data.

Christina

**Reply by JohnHalleyGotway** [reply]
@ZofiaBaldysz and @CPKalb, @hsoh-u is working on the a bugfix including the handling of this scale factor of 0.9.

I went hunting for a sample data file on our ftp site, but couldn't find one.

@ZofiaBaldysz can you please send us a sample data file we can use for testing? It can be [uploaded](https://github.com/dtcenter/METplus/discussions/954) here. And then respond here so we know to go look for it.

Thanks.

**Reply by ZofiaBaldysz** [reply]
@CPKalb and @JohnHalleyGotway I apologize for the delayed answer.
I've prepared some sample files but I will describe the problem again to understand my issue better.
In NETCDF CF documentation (https://cfconventions.org/cf-conventions/cf-conventions.html#polar-stereographic), there is information that polar stereographic projection requires the following attributes:
- straight_vertical_longitude_from_pole
- latitude_of_projection_origin - Either +90. or -90.
- Either standard_parallel (EPSG 9829) or scale_factor_at_projection_origin (EPSG 9810)
- false_easting - This parameter is optional (default is 0)
- false_northing - This parameter is optional (default is 0)

First issue:
When I provided `straight_vertical_longitude_from_pole`,  `latitude_of_projection_origin` and `standard_parallel` and ran:
`plot_data_plane radolan.nc radolan.ps 'name="PREC"; level="(0,*,*)";'`
I obtained the error:
`get_nc_var_att_double() -> Cannot get "longitude_of_projection_origin" from RADOLAN_grid variable.`

Second issue:
To my previous case I added `longitude_of_projection_origin` and I got error:
`get_nc_var_att_double() -> Cannot get "scale_factor_at_projection_origin" from RADOLAN_grid variable.`

Third issue:
To my second issue I added `scale_factor_at_projection_origin` equal 0.93301270189 and I got:
`NcCfFile::get_grid_mapping_polar_stereographic() -> Unexpected attribute value of 0.933013 for the scale_factor_at_projection_origin attribute of the RADOLAN_grid variable.`
but in NETCDF CF documentation, we can find that the scale factor can be > 0.0

Fourth issue:
I removed `standard_parallel` and left only `scale_factor_at_projection_origin` and I obtained the same error:
`NcCfFile::get_grid_mapping_polar_stereographic() -> Unexpected attribute value of 0.933013 for the scale_factor_at_projection_origin attribute of the RADOLAN_grid variable.`

I've prepared two example files:
radolan_sp.nc - issue two (only standard_parallel)
radolan_sp_sf.nc - issue three (standard_parallel and scale_factor_at_projection_origin)
I will try to upload them to your ftp server.


Thanks,
Zofia

**Comment by hsoh-u** [comment]
The uploaded files are not recognized by MET and ncdump. Would you convert it to HDF5 or NetCDF files?

**Reply by ZofiaBaldysz** [reply]
The files are already in NetCDF. Maybe something went wrong with data upload. I've uploaded them once again. 

```ncdump -h radolan_sf.nc
netcdf radolan_sf {
dimensions:
        time = 1 ;
        y = 1100 ;
        x = 900 ;
variables:
        int time(time) ;
                time:standard_name = "time" ;
                time:long_name = "time minutes since 2020-01-01T00:00:00.000000000" ;
                time:units = "minutes since 2020-01-01T00:00:00.000000000" ;
        int RADOLAN_grid ;
                RADOLAN_grid:grid_mapping_name = "polar_stereographic" ;
                RADOLAN_grid:projection = "polar_stereographic" ;
                RADOLAN_grid:straight_vertical_longitude_from_pole = 10 ;
                RADOLAN_grid:latitude_of_projection_origin = 90 ;
                RADOLAN_grid:longitude_of_projection_origin = 10 ;
                RADOLAN_grid:scale_factor_at_projection_origin = 0.93301270189 ;
                RADOLAN_grid:false_easting = 0 ;
                RADOLAN_grid:false_northing = 0 ;
                RADOLAN_grid:semi_major_axis = 6370040 ;
                RADOLAN_grid:semi_minor_axis = 6370040 ;
        double lat(y, x) ;
                lat:standard_name = "latitude" ;
                lat:long_name = "latitude" ;
                lat:units = "degrees_north" ;
                lat:coorindates = "lat lon" ;
        double lon(y, x) ;
                lon:standard_name = "longitude" ;
                lon:long_name = "longitude" ;
                lon:units = "degrees_east" ;
                lon:coordinates = "lat lon" ;
        double y(y) ;
                y:standard_name = "projection_y_coordinate" ;
                y:long_name = "RADOLAN Grid y coordinate of projection" ;
                y:units = "km" ;
        double x(x) ;
                x:standard_name = "projection_x_coordinate" ;
                x:long_name = "RADOLAN Grid y coordinate of projection" ;
                x:units = "km" ;
        float PREC(time, y, x) ;
                PREC:_FillValue = NaNf ;
                PREC:standard_name = "rainfall_amount" ;
                PREC:long_name = "rainfall amount" ;
                PREC:description = "accumulated rainfall amount" ;
                PREC:units = "kg m-2" ;
                PREC:grid_mapping = "RADOLAN_grid" ;
                PREC:coordinates = "lat lon" ;

// global attributes:
                :TITLE = "Radar rainfall" ;
                :Conventions = "CF-1.7" ;
}
```

**Comment by hsoh-u** [comment]
Thank you, new files are good

**Comment by ZofiaBaldysz** [comment]
Hi!
Just wanted to check if there is any news regarding to our polar stereographic issue?
Thanks,
Zofia

**Comment by hsoh-u** [comment]
We made some progress and we aim to release at the next beta.

**Comment by hsoh-u** [comment]
The scale_factor is computed from the standard_parallel. Plot By MET (will be available with beta 5):
![polar_met](https://user-images.githubusercontent.com/15203106/202553219-c9a6c884-b109-4258-9466-8eb8061862c8.png)
Plot By other tool:
![polar_idv](https://user-images.githubusercontent.com/15203106/202553262-29ad2cca-1ace-44f2-97fd-8d59482d356a.png)

**Comment by hsoh-u** [comment, marked as answer]
This was not included to beta5. It needs more tests because there are two types of polar stereographics (spherical earth and ellipsoidal earth). MET only supports the spherical earth. The data for NSIDC Sea Ice Edge is the ellipsoidal earth
```
        char crs ;
                crs:grid_mapping_name = "polar_stereographic" ;
                crs:straight_vertical_longitude_from_pole = -45. ;
                crs:false_easting = 0. ;
                crs:false_northing = 0. ;
                crs:latitude_of_projection_origin = 90. ;
                crs:standard_parallel = 70. ;
                crs:long_name = "NSIDC_NH_PolarStereo_25km" ;
                crs:longitude_of_prime_meridian = 0. ;
                crs:semi_major_axis = 6378273. ;
                crs:inverse_flattening = 298.279411123064 ;
```

**Reply by hsoh-u** [reply]
We aim for the MET v11.0.1 bugfix release.
Regridding a spherical earth with the parallel standard instead of scale factor (input: radolan_sp.nc):
![polar_RADORAN_regrid](https://user-images.githubusercontent.com/15203106/215817517-fdb3d728-b54b-4f48-824b-69fc7d012ca0.png)

Regridding an ellipsoidal earth with the parallel standard instead of scale factor and additional attributes (input: NSIDC0081_SEAICE_PS_N25km_20220723_v2.0.nc):
```
                crs:standard_parallel = 70. ;
                crs:false_easting = 0. ;
                crs:false_northing = 0. ;
                ...
                crs:semi_major_axis = 6378273. ;
                crs:inverse_flattening = 298.279411123064 ;
```
![polar_ellipsosidal_north_regrid](https://user-images.githubusercontent.com/15203106/215817985-305af9cc-7b2c-42dc-96ad-25feaafff148.png)

**Reply by JohnHalleyGotway** [reply]
Marking as answered since the fix was provided in the 11.0.1 bugfix release.
